File: C#/Leetcode/Patterns/Intervals/EmployeeFreeTime.cs
using System;
using System.Collections.Generic;

namespace Leetcode.Patterns.Intervals {
    // Leetcode 759 - https://leetcode.com/problems/employee-free-time/
    // Tx = O(n*logn)
    // Sx = O(n)
    public class EmployeeFreeTime {
        // Step 1: Combine all the intervals of all the employees.
        // Step 2: Sort the intervals based on the start time.
        // Step 3: Merge the intervals.
        // Step 4: Find the free time between the merged intervals which is the combination of end of one interval and start of the next interval.
        public int[][] EmployeeFreeTime(int[][] schedule) {
            // Step 1
            IList<int[]> combinedIntervals = new List<int[]>();
            foreach (var employee in schedule) {
                foreach (var interval in employee) {
                    combinedIntervals.Add(interval);
                }
            }
            
            // Step 2
            combinedIntervals.Sort((i1, i2) => i1[0].CompareTo(i2[0]));

            // Step 3: Merge Intervals
            IList<int[]> mergedIntervals = new List<int[]>();
            mergedIntervals.Add(combinedIntervals[0]);
            int mergedIndex = 0, intervalIndex = 1, length = combinedIntervals.GetLength(0);

            while(intervalIndex < length) {
                if(mergedIntervals[mergedIndex][1] >= combinedIntervals[intervalIndex][0]) {
                    mergedIntervals[mergedIndex][1] = Math.Max(mergedIntervals[mergedIndex][1], combinedIntervals[intervalIndex][1]);
                } else {
                    mergedIntervals.Add(combinedIntervals[intervalIndex]);
                    mergedIndex++;
                }
                intervalIndex++;
            }

            // Step 4
            List<int[]> result = new List<int[]>();
            for(int index = 1; index < mergedIntervals.Count; index++ ) {
                int[] start = mergedIntervals[index-1];
                int[] end = mergedIntervals[index];
                result.Add(new int[](start[1], end [0]));
            }

            return result.ToArray();
        }
    }
}

File: Java/Intervals/EmployeeFreeTime.java
package Intervals;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class EmployeeFreeTime {
    // Input: schedule = [[[2,4],[7,10]],[[1,5]],[[6,9]]] => [[1,5], [2,4], [6,9], [7,10]]
    // Output: [(5,6)]

    public int[][] FreeTime(int[][] schedule) {
        List<int[]> freeTime = new ArrayList<>();
        Arrays.sort(schedule, Comparator.comparingInt(s -> s[0]));

        // Step1: Merge the intervals;
        List<int[]> mergedIntervals = new ArrayList<>();
        mergedIntervals.add(schedule[0]);

        for(int index=1; index<schedule.length; index++) {
            int size = mergedIntervals.size()-1;
            if(schedule[index][0] <= mergedIntervals.get(size)[1]) {
                mergedIntervals.get(size)[1] = Math.max(mergedIntervals.get(size)[1], schedule[index][1]);
            } else {
                mergedIntervals.add(schedule[index]);
            }
        }

        // Step 2: Find the gap between each merged intervals which is the free time.
        for(int index = 1; index<mergedIntervals.size(); index++) {
            int start = mergedIntervals.get(index-1)[1];
            int end = mergedIntervals.get(index)[0];
            freeTime.add(new int[]{start, end});
        }

        return freeTime.toArray(new int[freeTime.size()][]);
    }
}
